File: visual-builder/src/types/schedule.ts
/**
 * Schedule type definitions for Visual Builder
 *
 * This module defines schedule structures for workflow automation
 * including cron-based schedules and their management.
 */

/**
 * Complete schedule definition
 *
 * Represents a scheduled trigger for a workflow execution.
 */
export interface Schedule {
  /** Unique identifier for this schedule */
  id: string;
  /** ID of the workflow this schedule triggers */
  workflowId: string;
  /** Schedule type - currently only cron is supported */
  type: 'cron';
  /** Cron expression (e.g., "0 9 * * *") */
  expression: string;
  /** Input data to pass to the workflow */
  input: Record<string, unknown>;
  /** Timezone for the schedule (e.g., "America/New_York") */
  timezone: string;
  /** Whether the schedule is currently active */
  enabled: boolean;
  /** ISO timestamp of last execution, null if never run */
  lastRunAt: string | null;
  /** ISO timestamp of next scheduled execution, null if disabled */
  nextRunAt: string | null;
  /** ISO timestamp of schedule creation */
  createdAt: string;
}

/**
 * Data for creating a new schedule
 */
export interface CreateScheduleData {
  /** ID of the workflow to schedule */
  workflowId: string;
  /** Cron expression */
  expression: string;
  /** Optional input data for the workflow */
  input?: string;
  /** Optional timezone (defaults to user's local timezone) */
  timezone?: string;
}

/**
 * Request payload for creating a schedule via API
 */
export interface ScheduleCreateRequest {
  workflowId: string;
  expression: string;
  input?: string;
  timezone?: string;
}

/**
 * Data for updating an existing schedule
 */
export interface UpdateScheduleData {
  /** New cron expression */
  expression?: string;
  /** New input data */
  input?: string;
  /** New timezone */
  timezone?: string;
  /** Enable/disable the schedule */
  enabled?: boolean;
}

File: visual-builder/backend/app/dtos/schedule.py
"""Schedule Pydantic schemas for API request/response.

Phase -1 Implementation:
- ScheduleCreate, ScheduleUpdate, ScheduleResponse
- ScheduleListResponse
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class ScheduleCreate(BaseModel):
    """Schedule creation request."""

    workflow_id: str = Field(
        ...,
        min_length=1,
        max_length=36,
        description="Workflow ID to schedule",
        alias="workflowId",
    )
    expression: str = Field(
        ...,
        min_length=1,
        max_length=255,
        description="Cron expression (e.g., '0 9 * * *' for 9am daily)",
    )
    input: dict[str, Any] = Field(
        default_factory=dict,
        description="Input to pass to workflow execution",
    )
    timezone: str = Field(
        default="UTC",
        max_length=50,
        description="Timezone for schedule (e.g., 'America/New_York')",
    )

    model_config = ConfigDict(
        populate_by_name=True,
        json_schema_extra={
            "example": {
                "workflow_id": "wf-123",
                "expression": "0 9 * * *",
                "input": {"param": "value"},
                "timezone": "UTC",
            }
        }
    )


class ScheduleUpdate(BaseModel):
    """Schedule update request."""

    expression: str | None = Field(
        None,
        min_length=1,
        max_length=255,
        description="Cron expression",
    )
    input: dict[str, Any] | None = Field(
        None,
        description="Input to pass to workflow execution",
    )
    timezone: str | None = Field(
        None,
        max_length=50,
        description="Timezone for schedule",
    )
    enabled: bool | None = Field(
        None,
        description="Whether schedule is enabled",
    )


class ScheduleResponse(BaseModel):
    """Schedule response."""

    id: str = Field(..., description="Schedule ID")
    workflow_id: str = Field(..., description="Workflow ID", alias="workflowId")
    type: str = Field(..., description="Schedule type (cron)")
    expression: str = Field(..., description="Cron expression")
    input: dict[str, Any] = Field(..., description="Workflow input")
    timezone: str = Field(..., description="Schedule timezone")
    enabled: bool = Field(..., description="Whether schedule is enabled")
    last_run_at: datetime | None = Field(None, description="Last execution time", alias="lastRunAt")
    next_run_at: datetime | None = Field(None, description="Next scheduled run", alias="nextRunAt")
    created_at: datetime = Field(..., description="Creation timestamp", alias="createdAt")

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class ScheduleListResponse(BaseModel):
    """Schedule list response."""

    schedules: list[ScheduleResponse] = Field(
        ...,
        description="List of schedules",
    )
    total: int = Field(..., description="Total count")
